qsc/plot.py

```diff
@@ -0,0 +1,529 @@
+        # Rotate surfaces
+        for th in range(x_2D_plot.shape[0]):
+            for ph in range(x_2D_plot.shape[1]):
+                [x_2D_plot_rotated[th,ph], y_2D_plot_rotated[th,ph], z_2D_plot_rotated[th,ph]] = rx.apply(rz.apply(np.array([x_2D_plot[th,ph], y_2D_plot[th,ph], z_2D_plot[th,ph]])))
+        # Rotate field lines
+        for th in range(fieldline_X.shape[0]):
+            for ph in range(fieldline_X.shape[1]):
+                [fieldline_X_rotated[th,ph], fieldline_Y_rotated[th,ph], fieldline_Z_rotated[th,ph]] = rx.apply(rz.apply(np.array([fieldline_X[th,ph], fieldline_Y[th,ph], fieldline_Z[th,ph]])))
+        # Plot surfaces
+        mlab.mesh(x_2D_plot_rotated, y_2D_plot_rotated-shift_array[i], z_2D_plot_rotated, scalars=Bmag, colormap='viridis')
+        # Plot field lines
+        for j in range(len(alphas)):
+            mlab.plot3d(fieldline_X_rotated[j], fieldline_Y_rotated[j]-shift_array[i], fieldline_Z_rotated[j], color=(0,0,0), line_width=0.001, tube_radius=0.005)
+
+def get_boundary(self, r=0.1, ntheta=40, nphi=130, ntheta_fourier=20, mpol=13, ntor=25):
+    '''
+    Function that, for a given near-axis radial coordinate r, outputs
+    the [X,Y,Z,R,Z] components of the boundary. The resolution along the toroidal
+    angle phi is equal to the resolution nphi for the axis, while ntheta
+    is specified by the used.
+
+    Args:
+      r (float): near-axis radius r where to create the surface
+      ntheta (int): Number of grid points to plot in the poloidal angle.
+      nphi   (int): Number of grid points to plot in the toroidal angle.
+      ntheta_fourier (int): Resolution in the Fourier transform to cylindrical coordinates
+      mpol: resolution in poloidal Fourier space
+      ntor: resolution in toroidal Fourier space
+    '''
+    # Get surface shape at fixed off-axis toroidal angle phi
+    R_2D, Z_2D, _ = self.Frenet_to_cylindrical(r, ntheta=ntheta_fourier)
+    # Get Fourier coefficients in order to plot with arbitrary resolution
+    RBC, RBS, ZBC, ZBS = to_Fourier(R_2D, Z_2D, self.nfp, mpol=mpol, ntor=ntor, lasym=self.lasym)
+    if not self.lasym:
+        RBS = np.zeros((int(2*ntor+1),int(mpol+1)))
+        ZBC = np.zeros((int(2*ntor+1),int(mpol+1)))
+
+    theta1D = np.linspace(0,2*np.pi,ntheta)
+    phi1D = np.linspace(0,2*np.pi,nphi)
+    phi2D, theta2D = np.meshgrid(phi1D,theta1D)
+    R_2Dnew = np.zeros((ntheta,nphi))
+    Z_2Dnew = np.zeros((ntheta,nphi))
+    for m in range(mpol+1):
+        for n in range(-ntor, ntor+1):
+            angle = m * theta2D - n * self.nfp * phi2D
+            R_2Dnew += RBC[n+ntor,m] * np.cos(angle) + RBS[n+ntor,m] * np.sin(angle)
+            Z_2Dnew += ZBC[n+ntor,m] * np.cos(angle) + ZBS[n+ntor,m] * np.sin(angle)
+
+    # X, Y, Z arrays for the whole surface
+    x_2D_plot = R_2Dnew*np.cos(phi1D)
+    y_2D_plot = R_2Dnew*np.sin(phi1D)
+    z_2D_plot = Z_2Dnew
+
+    return x_2D_plot, y_2D_plot, z_2D_plot, R_2Dnew, Z_2Dnew
```

Comment:
It seems odd to return z twice. Ok with you if only the 1st 4 arrays are returned?

Reply:
Sure. Fixed!